### DISCUSSION TITLE
RadiusGraph/KNNGraph based on node type

### DISCUSSION THREAD

**Discussion by niphiwi** [pyg-team/pytorch_geometric#7771] [General]
Hi, is there already something implemented to create edges based on node types? For example, I have "observed" and "unobserved" nodes and want to create undirected edges with RadiusGraph between observed nodes, but only directed edges from observed to unobserved.
I am currently fiddling around with creating multiple subgraphs and filtering them, but this seems overly complicated to me.

EDIT: I might need to clarify that I am working with a homogeneous graph, i.e., I use a boolean mask that identifies nodes as observed/unobserved.

**Comment by rusty1s** [comment]
I think what I would do is create a `RadiusGraph`, and then just remove any invalidate edge, e.g.
```python
observed_node_mask = ...  # Node-level mask to check whether a node is observed

edge_mask = observed_node_mask[edge_index[0]] & observed_node_mask[edge_index[1]]
edge_index = edge_index[:, edge_mask]
```

**Reply by niphiwi** [reply]
Thanks for your help! 😊 I did not think about that simple idea of an edge mask. Instead, I tried to filter the edge_index overly complicated with list comprehensions.
